coral-trino/src/main/java/com/linkedin/coral/trino/rel2trino/transformers/CastOperatorTransformer.java

```diff
@@ -0,0 +1,122 @@
+ * for queries with CAST operators, there is an added (but expected) side effect of implicit casting by Calcite's type coercion rules.
+ * Consider the following Hive input query:
+ * "SELECT CAST(from_utc_timestamp(a_date, 'America/Los_Angeles') AS DECIMAL(10, 0)) AS d FROM test.table_from_utc_timestamp"
+ *
+ * We add a `with_timezone` operator in this transformation, and with the added layer of Calcite validation, instead
+ * of just getting:
+ * "(with_timezone(table_from_utc_timestamp0.a_date), 'UTC')"
+ *
+ * We get an extra implicit cast from Calcite on the input to `with_timezone` which is a date type to the expected timestamp type:
+ * "(with_timezone(CAST(table_from_utc_timestamp0.a_date AS TIMESTAMP), 'UTC'))"
+ */
+public class CastOperatorTransformer extends SqlCallTransformer {
+  private static final String WITH_TIMEZONE = "with_timezone";
+  private static final String TO_UNIXTIME = "to_unixtime";
+  private static final Set<SqlTypeName> BINARY_SQL_TYPE_NAMES =
+      new HashSet<>(Arrays.asList(SqlTypeName.VARBINARY, SqlTypeName.BINARY));
+  private static final Set<SqlTypeName> CHAR_SQL_TYPE_NAMES =
+      new HashSet<>(Arrays.asList(SqlTypeName.VARCHAR, SqlTypeName.CHAR));
+
+  public CastOperatorTransformer(TypeDerivationUtil typeDerivationUtil) {
+    super(typeDerivationUtil);
+  }
+
+  @Override
+  protected boolean condition(SqlCall sqlCall) {
+    return sqlCall.getOperator().getKind() == SqlKind.CAST;
+  }
+
+  @Override
+  protected SqlCall transform(SqlCall sqlCall) {
+    List<SqlNode> operands = sqlCall.getOperandList();
+
+    SqlNode leftOperand = operands.get(0);
+    RelDataType sourceType = deriveRelDatatype(leftOperand);
+    SqlDataTypeSpec targetSqlDataTypeSpec = (SqlDataTypeSpec) operands.get(1);
+    SqlTypeName targetType = SqlTypeName.get(targetSqlDataTypeSpec.getTypeNameSpec().getTypeName().toString());
+
+    if (sourceType.getSqlTypeName() == TIMESTAMP && targetType == DECIMAL) {
+      SqlOperator trinoWithTimeZone =
+          createSqlOperator(WITH_TIMEZONE, explicit(TIMESTAMP /* should be WITH TIME ZONE */));
+      SqlOperator trinoToUnixTime = createSqlOperator(TO_UNIXTIME, explicit(DOUBLE));
+
+      SqlCall withTimeZoneCall =
+          trinoWithTimeZone.createCall(ZERO, leftOperand, CalciteUtil.createStringLiteral("UTC", ZERO));
+      SqlCall toUnixTimeCall = trinoToUnixTime.createCall(ZERO, withTimeZoneCall);
+
+      return castOperand(toUnixTimeCall, targetSqlDataTypeSpec);
+    }
+
+    if (BINARY_SQL_TYPE_NAMES.contains(sourceType.getSqlTypeName()) && CHAR_SQL_TYPE_NAMES.contains(targetType)) {
+      SqlOperator fromUTF8 = createSqlOperator("from_utf8", explicit(VARCHAR));
+
+      return fromUTF8.createCall(ZERO, leftOperand);
+    }
+
+    return sqlCall;
+  }
+
+  private SqlCall castOperand(SqlNode operand, SqlDataTypeSpec targetSqlDataTypeSpec) {
+    return SqlStdOperatorTable.CAST.createCall(ZERO, operand, targetSqlDataTypeSpec);
```

Comment:
This method is only used once, I don't think it's necessary to create a special helper method.

Reply:
Removed